compendium/ConfigurationAdmin/src/ThreadpoolSafeFuturePrivate.cpp

```diff
@@ -0,0 +1,69 @@
+/*=============================================================================
+
+  Library: CppMicroServices
+
+  Copyright (c) The CppMicroServices developers. See the COPYRIGHT
+  file at the top-level directory of this distribution and at
+  https://github.com/CppMicroServices/CppMicroServices/COPYRIGHT .
+
+  Licensed under the Apache License, Version 2.0 (the "License");
+  you may not use this file except in compliance with the License.
+  You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+
+=============================================================================*/
+
+#include "ThreadpoolSafeFuturePrivate.hpp"
+#include <chrono>
+#include <future>
+
+namespace cppmicroservices::cmimpl
+{
+
+    ThreadpoolSafeFuturePrivate::ThreadpoolSafeFuturePrivate(std::shared_future<void> future,
+                                                             std::shared_ptr<std::atomic<bool>> asyncStarted,
+                                                             std::shared_ptr<std::packaged_task<void(bool)>> task)
+        : future(future)
+        , asyncStarted(asyncStarted)
+        , task(task)
+    {
+    }
+    void
+    ThreadpoolSafeFuturePrivate::get() const
+    {
+        return wait();
+    }
+    void
+    ThreadpoolSafeFuturePrivate::wait() const
+    {
+        // ensure that asyncTaskMap and asyncTaskStateMap are cleared even if task throws
+
+        constexpr auto timeout = std::chrono::milliseconds(50);
+        // if we hit the timeout
+        if (future.wait_for(timeout) == std::future_status::timeout)
+        {
+            // we expect that the asyncStarted is false -- i.e. stalled
+            auto expected = false;
+            auto desired = true;
+            // if it is *asyncStarted==false
+            if (std::atomic_compare_exchange_strong(&(*asyncStarted), &expected, desired))
```

Comment:
technically no... In all the instances where a threadpoolsafefutureprivate is instantiated with a nullptr for asyncStarted, the future's value is set -- this means that `future.wait_fot(timeout)` above will always finish with `status != std::future_status::timeout`, therefore not getting to this check... That being said, as a point of forward compatibility if this ever changes (though I can't think of why it would) it is maybe a good idea to add this check in anyway?